snapcraft/plugins/_ros/ros2.py

```diff
@@ -0,0 +1,181 @@
+
+        packages = ['cmake']
+
+        # Dependencies for FastRTPS
+        packages.extend(['libasio-dev', 'libtinyxml2-dev'])
+
+        # Dependencies for the rest of ros2
+        packages.extend([
+            'libopencv-dev', 'libpoco-dev', 'libpocofoundation9v5',
+            'libpocofoundation9v5-dbg', 'python3-dev', 'python3-empy',
+            'python3-nose', 'python3-pip', 'python3-setuptools',
+            'python3-yaml', 'libtinyxml-dev', 'libeigen3-dev'])
+
+        return packages
+
+    def get_stage_packages(self):
+        """Return the packages required for running the underlay."""
+
+        # Ament is a python3 tool, and it requires setuptools at runtime
+        return ['python3', 'python3-setuptools']
+
+    def pull(self):
+        """Download the ROS2 underlay.
+
+        This first downloads the tools required to fetch the underlay, and then
+        uses those tools fetch the underlay. Both steps are run in a manner
+        such that they are only run once, even if subsequent steps fail.
+        """
+        self._run_step(
+            self._install_tools,
+            step=_INSTALL_TOOLS_STEP,
+            skip_message='Tools already installed. Skipping...')
+        self._run_step(
+            self._fetch_ros2,
+            step=_FETCH_ROS2_STEP,
+            skip_message='ros2 already fetched. Skipping...')
+
+    def build(self):
+        """Build the ROS2 underlay.
+
+        Build the ROS2 underlay that was fetched in pull(). This is only done
+        once, assuming success, which means calling this function multiple
+        times does not necessarily build multiple times.
+
+        :return: The path into which the underlay is installed.
+        :rtype: str
+        """
+        self._run_step(
+            self._build_ros2,
+            step=_BUILD_ROS2_STEP,
+            skip_message='ros2 already built. Skipping...')
+
+        return self._install_dir
+
+    def clean(self):
+        """Clean everything done in this class."""
+
+        with contextlib.suppress(FileNotFoundError):
+            shutil.rmtree(self._bootstrap_path)
+
+    def _run(self, command):
+        env = os.environ.copy()
+        env['PATH'] = env['PATH'] + ':' + os.path.join(
+            self._tool_dir, 'usr', 'bin')
+        env['PYTHONPATH'] = os.path.join(
+            self._tool_dir, 'usr', 'lib', 'python3', 'dist-packages')
+
+        subprocess.check_call(command, env=env)
+
+    def _step_done(self, step):
+        return os.path.isfile(os.path.join(self._state_dir, step))
+
+    def _set_step_done(self, step):
+        os.makedirs(self._state_dir, exist_ok=True)
+        open(os.path.join(self._state_dir, step), 'w').close()
+
+    def _run_step(self, callable, *, step, skip_message=None):
+        if self._step_done(step):
+            if skip_message:
+                logger.debug(skip_message)
+        else:
+            callable()
+            self._set_step_done(step)
+
+    def _install_tools(self):
```

Comment:
What do you think about _install_vcstool instead of _install_tools?

Reply:
It's too specific: this class needs to be able to grow as ros2 changes.